src/Validation/EditorConfigValidatorRules.cs

```diff
@@ -141,9 +143,191 @@ private void ValidateSections()
+                        EnforcementLevel = rule.EnforcementLevel,
+                    }).ToImmutableArray());
+
+            var reorderedOverlappingRules = new List<(string firstRule, string secondRule)>();
+            ImmutableArray<NamingRule> declaredNamingRules = namingStyle.Rules.NamingRules;
+            ImmutableArray<NamingRule> orderedNamingRules = orderedRulePreferences.Rules.NamingRules;
+            for (int i = 0; i < declaredNamingRules.Length; i++)
+            {
+                for (int j = 0; j < i; j++)
+                {
+                    NamingRule firstRule = declaredNamingRules[j];
+                    NamingRule secondRule = declaredNamingRules[i];
+
+                    // If the reordered rules are in the same relative order, no need to check the overlap
+                    bool reordered = false;
+                    foreach (NamingRule rule in orderedNamingRules)
+                    {
+                        if (rule.Name == firstRule.Name)
+                        {
+                            break;
+                        }
+                        else if (rule.Name == secondRule.Name)
+                        {
+                            reordered = true;
+                            break;
+                        }
+                    }
+
+                    if (!reordered)
+                    {
+                        continue;
+                    }
+
+                    // If the rules don't overlap, reordering is not a problem
+                    if (!Overlaps(firstRule, secondRule))
+                    {
+                        continue;
+                    }
+
+                    reorderedOverlappingRules.Add((firstRule.Name, secondRule.Name));
+                }
+            }
+
+            var reportedRules = new HashSet<string>();
+            foreach (Property property in section.Properties)
+            {
+                string name = property.Keyword.Text.Trim();
+                if (!name.StartsWith("dotnet_naming_rule."))
+                {
+                    continue;
+                }
+
+                string[] nameSplit = name.Split('.');
+                if (nameSplit.Length != 3)
+                {
+                    continue;
+                }
+
+                string ruleName = nameSplit[1];
+                if (!reportedRules.Add(ruleName))
+                {
+                    continue;
+                }
+
+                foreach ((string firstRule, string secondRule) in reorderedOverlappingRules)
+                {
+                    if (secondRule != ruleName)
+                    {
+                        continue;
+                    }
+
+                    ErrorCatalog.NamingRuleReordered.Run(property.Keyword, e =>
+                    {
+                        e.Register(secondRule, firstRule);
+                    });
+                }
+            }
+        }
+
+        private bool Overlaps(in NamingRule x, in NamingRule y)
```

Comment:
🐛 This implementation doesn't currently handle cases where the `ApplicableSymbolKindList` and `RequiredModifierList` are mutually-exclusive. For example, it thinks the following two rules overlap:

* Members with the `async` keyword are pascal case
* Fields and locals are camelCase

Since there are no symbols which match both of these rules, it does not matter what order they appear in in the .editorconfig file. Therefore, we should not produce a warning.

I didn't fix this bug in the first pass since it will take some thought to make sure other bugs are not introduced. We can address it as a follow-up. Very few .editorconfig files will be impacted by this.